modules/index/src/main/java/org/apache/ignite/internal/index/IndexTaskManager.java

```diff
@@ -0,0 +1,207 @@
+ *     <li>{@link CatalogEvent#INDEX_REMOVED} - when an index is removed from the Catalog, all ongoing tasks for the index will be stopped;
+ *     </li>
+ *     <li>{@link PrimaryReplicaEvent#PRIMARY_REPLICA_ELECTED} - if the local node has become the primary replica for partition {@code 0} of
+ *     the table, it starts corresponding tasks depending on the current status of indices in the Catalog. If the local node stops being the
+ *     primary replica for the partition {@code 0}, it stops all tasks that belong to indices of the table that the partition belongs to.
+ *     </li>
+ * </ul>
+ *
+ * <p>On node recovery, tasks will be started on {@link PrimaryReplicaEvent#PRIMARY_REPLICA_ELECTED}, which will fire due to a change in
+ * the {@link ReplicaMeta#getLeaseholderId()} on node restart but after {@link ReplicaMeta#getExpirationTime()}.
+ */
+class IndexTaskManager implements ManuallyCloseable {
+    private final CatalogService catalogService;
+
+    private final PlacementDriver placementDriver;
+
+    private final ClusterService clusterService;
+
+    private final ChangeIndexStatusTaskScheduler changeIndexStatusTaskScheduler;
+
+    /** Tables IDs for which the local node is the primary replica for the partition with ID {@code 0}. */
+    private final Set<Integer> localNodeIsPrimaryReplicaForTableIds = ConcurrentHashMap.newKeySet();
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    IndexTaskManager(
+            CatalogManager catalogManager,
+            PlacementDriver placementDriver,
+            ClusterService clusterService,
+            ChangeIndexStatusTaskScheduler changeIndexStatusTaskScheduler
+    ) {
+        this.catalogService = catalogManager;
+        this.placementDriver = placementDriver;
+        this.clusterService = clusterService;
+        this.changeIndexStatusTaskScheduler = changeIndexStatusTaskScheduler;
+
+        addListeners();
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        changeIndexStatusTaskScheduler.close();
+    }
+
+    private void addListeners() {
+        catalogService.listen(CatalogEvent.INDEX_CREATE, adaptToEventListener(this::onIndexCreated));
+
+        catalogService.listen(CatalogEvent.INDEX_STOPPING, adaptToEventListener(this::onIndexDropped));
+
+        catalogService.listen(CatalogEvent.INDEX_REMOVED, adaptToEventListener(this::onIndexRemoved));
+
+        placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, adaptToEventListener(this::onPrimaryReplicaElected));
+    }
+
+    /** Wraps a given callback into an EventListener. */
+    private <T extends EventParameters> EventListener<T> adaptToEventListener(Consumer<T> action) {
```

Comment:
Looks like a candidate for `IgniteUtils`.

Reply:
Definitely not in this PR. I'll have a look if this method can be reused in other places.

Reply:
I don’t insist, but maybe in this PR method we’ll add it to `IgniteUtils` and in another where we can use it?

Reply:
Created a static method in `EventListener`

Reply:
Commented on the new method.